### ISSUE TITLE
How to make it working for non UK numbers

### ISSUE THREAD

**Issue by abhishekraghav** [nickpack/keycloak-sms-authenticator#3]
Hey,
First of all thanks for writing this custom authenticator. 
I tried setting up the MFA for one of my realm. But i am getting below error when i am trying to add a non UK number after the user is registered. Does it work only for UK based numbers or can i tweak it to work for non UK as well. 
Thanks.

![image](https://user-images.githubusercontent.com/19988460/32596626-f707353c-c559-11e7-8ca1-c46cec490912.png)

**Comment by nickpack** [comment]
Hi, currently the validator is purely for UK numbers as this was my use case, the validator is here https://github.com/nickpack/keycloak-sms-authenticator-sns/blob/develop/src/main/java/six/six/keycloak/authenticator/KeycloakSmsAuthenticatorUtil.java#L114

If I get a chance this week I will make this configurable.

**Comment by abhishekraghav** [comment]
Thanks a lot. 👍

**Comment by abhishekraghav** [comment]
I am thing i have observed that once i configure the authenticator in my keycloak realm and then after sometime if i try opening any user by clicking on the userID link like as shown below, 

![image](https://user-images.githubusercontent.com/19988460/32609750-b0c74c38-c586-11e7-99a0-3d2880eacc2d.png)

I get the below exception stack trace.  FYI i am using 3.1.0.FInal version of keycloak here and I have checked out your develop branch to build. I have not changed anything in the pom.xml

```
19:41:00,154 ERROR [io.undertow.request] (default task-46) UT005023: Exception handling request to /auth/admin/realms/MFA-SMS/users/09001e9e-364e-41ee-a44d-805c8d2666c6: org.jboss.resteasy.spi.UnhandledException: java.lang.NoSuchMethodError: org.keycloak.models.cache.CachedUserModel.getCachedWith()Ljava/util/concurrent/ConcurrentMap;
        at org.jboss.resteasy.core.ExceptionHandler.handleApplicationException(ExceptionHandler.java:76)
        at org.jboss.resteasy.core.ExceptionHandler.handleException(ExceptionHandler.java:212)
        at org.jboss.resteasy.core.SynchronousDispatcher.writeException(SynchronousDispatcher.java:168)
        at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:411)
        at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:202)
        at org.jboss.resteasy.plugins.server.servlet.ServletContainerDispatcher.service(ServletContainerDispatcher.java:221)
        at org.jboss.resteasy.plugins.server.servlet.HttpServletDispatcher.service(HttpServletDispatcher.java:56)
        at org.jboss.resteasy.plugins.server.servlet.HttpServletDispatcher.service(HttpServletDispatcher.java:51)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
        at io.undertow.servlet.handlers.ServletHandler.handleRequest(ServletHandler.java:85)
        at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:129)
        at org.keycloak.services.filters.KeycloakSessionServletFilter.doFilter(KeycloakSessionServletFilter.java:90)
        at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:60)
        at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
        at io.undertow.servlet.handlers.FilterHandler.handleRequest(FilterHandler.java:84)
        at io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:62)
        at io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36)
        at org.wildfly.extension.undertow.security.SecurityContextAssociationHandler.handleRequest(SecurityContextAssociationHandler.java:78)
        at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
        at io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:131)
        at io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:57)
        at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
        at io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:46)
        at io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:64)
        at io.undertow.security.handlers.AuthenticationMechanismsHandler.handleRequest(AuthenticationMechanismsHandler.java:60)
        at io.undertow.servlet.handlers.security.CachedAuthenticatedSessionHandler.handleRequest(CachedAuthenticatedSessionHandler.java:77)
        at io.undertow.security.handlers.NotificationReceiverHandler.handleRequest(NotificationReceiverHandler.java:50)
        at io.undertow.security.handlers.AbstractSecurityContextAssociationHandler.handleRequest(AbstractSecurityContextAssociationHandler.java:43)
        at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
        at org.wildfly.extension.undertow.security.jacc.JACCContextIdHandler.handleRequest(JACCContextIdHandler.java:61)
        at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
        at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
        at io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:284)
        at io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:263)
        at io.undertow.servlet.handlers.ServletInitialHandler.access$000(ServletInitialHandler.java:81)
        at io.undertow.servlet.handlers.ServletInitialHandler$1.handleRequest(ServletInitialHandler.java:174)
        at io.undertow.server.Connectors.executeRootHandler(Connectors.java:202)
        at io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:793)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.NoSuchMethodError: org.keycloak.models.cache.CachedUserModel.getCachedWith()Ljava/util/concurrent/ConcurrentMap;
        at six.six.keycloak.authenticator.KeycloakSmsAuthenticatorCredentialProvider.onCache(KeycloakSmsAuthenticatorCredentialProvider.java:107)
        at org.keycloak.credential.UserCredentialStoreManager.onCache(UserCredentialStoreManager.java:302)
        at org.keycloak.models.cache.infinispan.UserCacheSession.onCache(UserCacheSession.java:413)
        at org.keycloak.models.cache.infinispan.UserCacheSession.cacheUser(UserCacheSession.java:373)
        at org.keycloak.models.cache.infinispan.UserCacheSession.getUserById(UserCacheSession.java:199)
        at org.keycloak.services.resources.admin.UsersResource.getUser(UsersResource.java:296)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:139)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:295)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:249)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:138)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:107)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:133)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:107)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:133)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:101)
        at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:395)
        ... 37 more

```

**Comment by abhishekraghav** [comment]
The above issue is fixed. I simply had to upgrade to 3.2.0.Final from 3.1.0.Final version of keycloak. Thanks.

**Comment by Giovanni-94** [comment]
Hi @nickpack, thanks  for writing this authenticator.
I wanted to ask if you had made the validator configurable. In my case, I need to configure the numbers of Italian mobiles.
Thanks in  advance

**Comment by Allan-Nava** [comment]
Hi @nickpack ,

is there a possibility to customize the configuration?

Thanks

**Comment by Giovanni-94** [comment]
Any news @nickpack?

**Comment by nickpack** [comment]
Currently still needs to be edited in the code, sorry - haven't had time to look at this yet.

**Comment by Allan-Nava** [comment]
@nickpack Ok, if u have news, let me know

**Comment by Allan-Nava** [comment]
@nickpack Is possible extend the base flow with add the mobile_number as user attribute with send sms code function?

**Comment by nickpack** [comment]
This is fixed by #8 - I need to review the changes before I merge, but looks good.

**Comment by nickpack** [comment]
Closing this as the translation files now allow you to set a regex per localisation. Thanks to @malys
